fixed bug in CenteredNorm, issue #19972

jpmattern opened the pull request:
## PR Summary

Fixed the CenteredNorm bug in issue #19972

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->

- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

smartlixx commented:
A test is needed IMHO.

dstansby reviewed (dismissed):
Yes, we should add a test here - there is code in the original issue (https://github.com/matplotlib/matplotlib/issues/19972) that should be fine.

tacaswell commented:
:+1: This can go in on one review once a test is aded.

jpmattern commented:
Sorry, I was away for the weekend. Happy to add a test. Basically one that recreates the circumstances of the reported bug, I presume?

jpmattern commented:
While the reported bug (#19972) raised an exception, the underlying issue was that the coded tested for `self.vmax is None` rather than `self._halfrange is None`. So in the old code, a manually set `halfrange` would have been reset by a call to `autoscale_None` would have reset it. Fixed that bug and added a small test that the old code would not have passed for both reasons: the exception and the reset of `autoscale_None`.

timhoffm commented:
Ping @dstansby

tacaswell commented:
@meeseeksbot backport to v3.4.x

QuLogic commented:
@meeseeksdev backport to v3.4.x
